crates/wasm-encoder/src/reencode.rs

```diff
@@ -806,6 +1226,127 @@ pub mod utils {
+    pub fn component_val_type<T: ?Sized + Reencode>(
+        reencoder: &mut T,
+        ty: wasmparser::ComponentValType,
+    ) -> crate::ComponentValType {
+        match ty {
+            wasmparser::ComponentValType::Primitive(pvt) => {
+                crate::ComponentValType::Primitive(reencoder.component_primitive_val_type(pvt))
+            }
+            wasmparser::ComponentValType::Type(index) => crate::ComponentValType::Type(index),
+        }
+    }
+
+    pub fn component_type_ref<T: ?Sized + Reencode>(
+        reencoder: &mut T,
+        ty: wasmparser::ComponentTypeRef,
+    ) -> crate::ComponentTypeRef {
+        match ty {
+            wasmparser::ComponentTypeRef::Module(index) => crate::ComponentTypeRef::Module(index),
+            wasmparser::ComponentTypeRef::Func(index) => crate::ComponentTypeRef::Func(index),
+            wasmparser::ComponentTypeRef::Value(vt) => {
+                crate::ComponentTypeRef::Value(reencoder.component_val_type(vt))
+            }
+            wasmparser::ComponentTypeRef::Type(bounds) => {
+                crate::ComponentTypeRef::Type(reencoder.type_bounds(bounds))
+            }
+            wasmparser::ComponentTypeRef::Instance(index) => {
+                crate::ComponentTypeRef::Instance(index)
+            }
+            wasmparser::ComponentTypeRef::Component(index) => {
+                crate::ComponentTypeRef::Component(index)
+            }
+        }
+    }
+
+    pub fn type_bounds<T: ?Sized + Reencode>(
+        _reencoder: &mut T,
+        bounds: wasmparser::TypeBounds,
+    ) -> crate::TypeBounds {
+        match bounds {
+            wasmparser::TypeBounds::Eq(index) => crate::TypeBounds::Eq(index),
+            wasmparser::TypeBounds::SubResource => crate::TypeBounds::SubResource,
+        }
+    }
+
+    pub fn component_export_kind<T: ?Sized + Reencode>(
+        _reencoder: &mut T,
+        kind: wasmparser::ComponentExternalKind,
+    ) -> crate::ComponentExportKind {
+        match kind {
+            wasmparser::ComponentExternalKind::Module => crate::ComponentExportKind::Module,
+            wasmparser::ComponentExternalKind::Func => crate::ComponentExportKind::Func,
+            wasmparser::ComponentExternalKind::Value => crate::ComponentExportKind::Value,
+            wasmparser::ComponentExternalKind::Type => crate::ComponentExportKind::Type,
+            wasmparser::ComponentExternalKind::Instance => crate::ComponentExportKind::Instance,
+            wasmparser::ComponentExternalKind::Component => crate::ComponentExportKind::Component,
+        }
+    }
+
+    pub fn alias<'a, T: ?Sized + Reencode>(
+        reencoder: &mut T,
+        alias: wasmparser::ComponentAlias<'a>,
+    ) -> crate::Alias<'a> {
+        match alias {
+            wasmparser::ComponentAlias::CoreInstanceExport {
+                kind,
+                instance_index,
+                name,
+            } => crate::Alias::CoreInstanceExport {
+                instance: instance_index,
```

Comment:
I had some similar comments on https://github.com/bytecodealliance/wasm-tools/pull/1655, but for indices like this could you also add a methods such as `component_instance_index` which default is a noop but allows the reencode trait to remap the indices if desired?

Reply:
How sophisticated should this be?  For example, in the case of `wasmparser::ComponentAlias::Outer`, the index seems like it refers to a different type of item depending on `kind` (and, per `count`, in a different scope, if that would affect the transformation) - should I therefore pass `kind` and `count` into the remapping function?  Sorry, I'm flying a bit blind here in terms of familiarity with the underlying concepts...!

Another example is cases such as `TypeBounds` where we are dealing with enums of indexes.  E.g. for TypeBounds the reencode logic is:

```
        match bounds {
            wasmparser::TypeBounds::Eq(index) => crate::TypeBounds::Eq(index),
            wasmparser::TypeBounds::SubResource => crate::TypeBounds::SubResource,
        }
```

Do I need to offer a separate hook to remap the index in cases like this, or would a consumer override the entire logic?